File: cpp/graph/bfs.cpp
#include <iostream>

#include "bfs.h"

void bfsSearch(Graph const &graph, int s, vector<int> &dist,
               vector<int> &pred) {
  const int n = graph.numVertices();
  pred.assign(n, -1);
  dist.assign(n, numeric_limits<int>::max());

  vector<vertexColor> color(n, White);

  dist[s] = 0;
  color[s] = Gray;

  queue<int> q;
  q.push(s);

  while (!q.empty()) {
    int u = q.front();

    for (VertexList::const_iterator iter = graph.begin(u); iter != graph.end(u);
         ++iter) {
      int v = iter->first;
      if (color[v] == White) {
        dist[v] = dist[u] + 1;
        pred[v] = u;
        color[v] = Gray;
        q.push(v);
      }
    }

    q.pop();
    color[u] = Black;
  }
}

File: dart/lib/graph/bfs.dart
import 'dart:collection';

import 'graph.dart';
import 'color.dart';

const int64Max = 9223372036854775807;

List<int> bfsSearch(Graph graph, int start) {
  final n = graph.n;
  final dist = List.generate(n, (_) => int64Max);
  final pred = List.generate(n, (_) => -1);
  final colors = List.generate(n, (_) => Color.white);

  dist[start] = 0;
  colors[start] = Color.gray;

  final q = Queue<int>();
  q.add(start);

  while (q.isNotEmpty) {
    final u = q.removeLast();
    for (var pair in graph.getVertices(u)) {
      final v = pair.first;
      if (colors[v] == Color.white) {
        dist[v] = dist[u] + 1;
        pred[v] = u;
        colors[v] = Color.gray;
        q.add(v);
      }
    }
    q.removeLast();
    colors[u] = Color.black;
  }

  return pred;
}
